### PR TITLE
[image][ios] fix resizing animated images failing to preserve animation

### LINKED ISSUE #28183: [expo-image] Animated images don't animate at small sizes
### Minimal reproducible example

https://github.com/bensaine/expo-gif-bug-mre

### What platform(s) does this occur on?

iOS

### Did you reproduce this issue in a development build?

Yes

### Summary

GIFs below a certain size will not animate on IOS, even if animation is started explicitly. See very simple repro. This 'cutoff' size seems to vary from GIF to GIF and device to device, but is a very problematic bug.

```js
export default function App() {
	const GIF_URL = "https://media1.tenor.com/m/ba6u_c1Mcb0AAAAC/futurenostlgia-laughing.gif"

	return (
		<View style={styles.container}>
			<Text>Normal remote GIF</Text>
			<Image autoplay={true} width={170} height={170} source={GIF_URL} />
			<Text>Same GIF at a smaller size (broken)</Text>
			<Image autoplay={true} width={150} height={150} source={GIF_URL} />
			<Text>Same GIF at a smaller size, local source (broken)</Text>
			<Image autoplay={true} width={150} height={150} source={require("./assets/tenor.gif")} />
		</View>
	)
}
```

https://github.com/expo/expo/assets/34781348/a436cd89-adf5-48c0-b9e2-e6cdfa45bf69

### Environment

expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 14.4.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 21.6.1 - /opt/homebrew/bin/node
      Yarn: 1.22.19 - /opt/homebrew/bin/yarn
      npm: 10.2.4 - /opt/homebrew/bin/npm
    Managers:
      CocoaPods: 1.14.3 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.4, iOS 17.4, macOS 14.4, tvOS 17.4, visionOS 1.1, watchOS 10.4
    IDEs:
      Android Studio: 2022.3 AI-223.8836.35.2231.11005911
      Xcode: 15.3/15E204a - /usr/bin/xcodebuild
    npmPackages:
      expo: ~50.0.14 => 50.0.15 
      react: 18.2.0 => 18.2.0 
      react-native: 0.73.6 => 0.73.6 
    npmGlobalPackages:
      eas-cli: 7.8.0
      expo-cli: 6.3.10
    Expo Workflow: managed

### Expo Doctor Diagnostics

✔ Check Expo config for common issues
✔ Check package.json for common issues
✔ Check native tooling versions
✔ Check dependencies for packages that should not be installed directly
✔ Check for common project setup issues
✔ Check for issues with metro config
✔ Check npm/ yarn versions
✔ Check Expo config (app.json/ app.config.js) schema
✔ Check for legacy global CLI installed locally
✔ Check that native modules do not use incompatible support packages
✔ Check that native modules use compatible support package versions for installed Expo SDK
✔ Check that packages match versions required by installed Expo SDK

Didn't find any issues with the project!

### PR DESCRIPTION
**Pull request by fobos531**
# Why

Fixes https://github.com/expo/expo/issues/28183

# How

### Research:

When dealing with animated images (GIFs), `expo-image` uses a `createAnimatedIfNeeded` method to properly prepare the GIF for the display. While this is all good, the problem arises in the fact that while a `SDAnimatedImage` is in fact a subclass of `UIImage`, it looks like the `SDAnimatedImage` constructor completely "loses" the `images` information on the resulting `UIImage`, which is used by standard `UIImage` to determine whether the image is animated or not. This means that in 

`func resize(animatedImage image: UIImage, toSize size: CGSize, scale: Double) async -> UIImage` 

the check for `image.images` is essentially invalid since it is guaranteed that this member will ALWAYS be `nil`. The reason this issue was not discovered before is that when `processImage` is run, downscaling will not aways run, depending on user's prop configuration vs. defaults set by `expo-image` in terms of whether the image should be downscaled or not. In events when the downscaling is indeed run, the resizing logic is also run, which means that the check for `image.images` will fail, and the method will return a still image since the that's the fallback in that specific method. When downscaling does **not** run, even though `SDAnimatedImage` constructor "lost" data about `image.images` array, the image will still render correctly since it's still an `SDAnimatedImage` in the end and no additional processing is run, which means it is unlikely that a still `UIImage` will be returned.

### Solution:

Use SDWebImage's `sd_isAnimated` extension property on `UIImage` to check whether the image is animated within the `func resize(animatedImage image: UIImage, toSize size: CGSize, scale: Double) async -> UIImage` function. We can use this since AFAIK all code paths into this function will ensure we're dealing with an instance of a SDWebImage or some of its subclasses. If the image is animated, cast it to `SDAnimatedImage` and use its instance methods/properties to perform the frame resizing logic. 

Error recovery mechanisms:
- image not animated -> return still image (resized)
- image animated but some frames failed to resize -> return animated image but with only the frames that were successfully resized
- image animated but NONE of the frames got resized -> return still image (resized)
- image animated, some other error happened with resizing - return still image (resized)

I'm not sure whether there are any negative performance implications to my implementation, but it seems to work. Happy to adjust it if you have any feedback.

# Test Plan

See middle and right image in first row - List of GIFs screen in NCL

Before:

https://github.com/expo/expo/assets/19533289/b2cdf022-2998-4af4-9f46-ec91053e8c80

After:

https://github.com/expo/expo/assets/19533289/316ccc57-99e5-456e-9f17-e566f7154355

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [X] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-image/ios/ImageUtils.swift [hunk 1]
```diff
@@ -116,19 +116,42 @@ func shouldDownscale(image: UIImage, toSize size: CGSize, scale: Double) -> Bool
...
+  if !image.sd_isAnimated {
+    return resize(image: image, toSize: size, scale: scale)
+  }
+
+  // Cast to SDAnimatedImage to access its properties
+  guard let image = image as? SDAnimatedImage else {
+    return resize(image: image, toSize: size, scale: scale)
+  }
+
+  let frameCount = await image.animatedImageFrameCount
```
Comment (tsapeta) [inline packages/expo-image/ios/ImageUtils.swift:128]: I guess we don't need `await` here 🤔 `animatedImageFrameCount` is just a member and doesn't take a closure as an argument
```suggestion
  let frameCount = image.animatedImageFrameCount
```

File: packages/expo-image/ios/ImageUtils.swift [hunk 2]
```diff
@@ -116,19 +116,42 @@ func shouldDownscale(image: UIImage, toSize size: CGSize, scale: Double) -> Bool
...
     // `animatedImage(with:duration:)` can return `nil`, probably when scales are not the same
-    // so it should never happen in our case, but let's handle it gracefully.
-    if let animatedImage = UIImage.animatedImage(with: resizedImages, duration: image.duration) {
-      return animatedImage
+    // so it should never happen in our case, but let's make sure to handle it gracefully.
+    if let newAnimatedImage = UIImage.animatedImage(with: resizedFrames, duration: image.duration) {
+      return newAnimatedImage
     }
+  } catch {
+    print("Error during concurrent image resizing: \(error)")
```
Comment (tsapeta) [inline packages/expo-image/ios/ImageUtils.swift:152]: Let's use our own logger for this
```suggestion
    log.error("Error during concurrent image resizing: \(error)")
```

File: packages/expo-image/ios/ImageUtils.swift [hunk 3]
```diff
@@ -116,19 +116,42 @@ func shouldDownscale(image: UIImage, toSize size: CGSize, scale: Double) -> Bool
...
+  // Cast to SDAnimatedImage to access its properties
+  guard let image = image as? SDAnimatedImage else {
+    return resize(image: image, toSize: size, scale: scale)
+  }
+
+  let frameCount = await image.animatedImageFrameCount
+
+  do {
+    let resizedFrames = try await concurrentMap(0..<frameCount) { index -> UIImage? in
+      guard let frame = await image.animatedImageFrame(at: index) else {
```
Comment (tsapeta) [inline packages/expo-image/ios/ImageUtils.swift:132]: Unnecessary `await`
```suggestion
      guard let frame = image.animatedImageFrame(at: index) else {
```

File: packages/expo-image/ios/ImageUtils.swift [hunk 4]
```diff
@@ -116,19 +116,42 @@ func shouldDownscale(image: UIImage, toSize size: CGSize, scale: Double) -> Bool
...
+        print("Warning: Frame at index \(index) could not be loaded and will be skipped.")
```
Comment (tsapeta) [inline packages/expo-image/ios/ImageUtils.swift:133]:
```suggestion
        log.warn("Frame at index \(index) could not be loaded and will be skipped.")
```

File: packages/expo-image/ios/ImageUtils.swift [hunk 5]
```diff
@@ -116,19 +116,42 @@ func shouldDownscale(image: UIImage, toSize size: CGSize, scale: Double) -> Bool
...
+        return nil
+      }
+      return resize(image: frame, toSize: size, scale: scale)
+    }.compactMap { $0 }
+
+    // Check if all frames were skipped or failed to load
+    if resizedFrames.isEmpty {
+      print("Error: All frames failed to load and/or were skipped.")
```
Comment (tsapeta) [inline packages/expo-image/ios/ImageUtils.swift:141]:
```suggestion
      log.error("All frames failed to load and/or were skipped.")
```

**Review by tsapeta — CHANGES_REQUESTED** [review]
Thanks for the detailed description and the fix! I added a few comments.

File: packages/expo-image/ios/ImageUtils.swift (hunk 1 above)
Comment (fobos531) [reply packages/expo-image/ios/ImageUtils.swift:128]: @tsapeta That's what I thought as well at first, but if you check autocomplete:

CleanShot 2024-04-21 at 18 51 22@2x

And you get a compiler warning:

CleanShot 2024-04-21 at 18 55 23@2x


Swift apparently has async computed properties: https://www.hackingwithswift.com/quick-start/concurrency/how-to-create-and-use-async-properties and this is what is the case with these SDAnimatedImage properties. Same goes for image.animatedImageFrame

CleanShot 2024-04-21 at 18 54 55@2x

**Comment by fobos531** [comment]
The suggestions have been applied. Please let me know if there's anything else to be done.

File: packages/expo-image/ios/ImageUtils.swift (hunk 3 above)
Comment (fobos531) [reply packages/expo-image/ios/ImageUtils.swift:132]: See https://github.com/expo/expo/pull/28335#discussion_r1573854867

**Review by lukmccall — APPROVED** [review]
Thanks for your contribution! Looks good, but let's wait for the final approval from @tsapeta

File: packages/expo-image/ios/ImageUtils.swift (hunk 1 above)
Comment (tsapeta) [reply packages/expo-image/ios/ImageUtils.swift:128]: That's interesting. Looks like all Objective-C properties are always async within the async context 🤔 I would rather expect them to be always sync 😅 TIL

**Review by tsapeta — APPROVED** [review]

**Comment by bensaine** [comment]
This seems to include a regression ! See https://github.com/expo/expo/issues/28183#issuecomment-2068200612

**Comment by FelipeSSantos1** [comment]
is this related to this issue? https://github.com/expo/expo/issues/25920#issuecomment-1944900358

**Pull request merged**